mbe: Clean up code with non-optional `NonterminalKind`

tgross35 opened the pull request:
Since [rust-lang/rust#128425], the fragment specifier is unconditionally required in all
editions. This means `NonTerminalKind` no longer needs to be optional,
as we can reject this code during the expansion of `macro_rules!` rather
than handling it throughout the code. Do this cleanup here.

[rust-lang/rust#128425]: https://github.com/rust-lang/rust/pull/128425

<!-- homu-ignore:start -->
r? @petrochenkov 
<!-- homu-ignore:end -->

tgross35 committed:
mbe: Clean up code with non-optional `NonterminalKind`

Since [1], the fragment specifier is unconditionally required in all
editions. This means `NonTerminalKind` no longer needs to be optional,
as we can reject this code during the expansion of `macro_rules!` rather
than handling it throughout the code. Do this cleanup here.

[1]: https://github.com/rust-lang/rust/pull/128425

tgross35 committed:
mbe: Unnest nested matches with let-else and let chains

Non-functional change to simplify control flow.

tgross35 commented:
Can be reviewed by-commit, the first commit is just refactoring a heavily nested function.

tgross35 commented on tests/ui/macros/macro-missing-fragment-deduplication.rs:
```diff
@@ -2,12 +2,11 @@
-    m!();
-    m!();
-    m!();
-    m!();
+    m!(); //~ ERROR unexpected end
+    m!(); //~ ERROR unexpected end
+    m!(); //~ ERROR unexpected end
+    m!(); //~ ERROR unexpected end
```
I don't know why these new "unexpected end" errors are showing up at the invocation rather than at `macro_rules!` or not at all. Any idea how to get rid of this?

petrochenkov replied:
> I don't know why these new "unexpected end" errors are showing up at the invocation rather than at `macro_rules!` or not at all. Any idea how to get rid of this?

Well, `m` now expects a `tt` in the input due to recovery, but `m!()` do not pass a `tt`.

tgross35 replied:
That all makes sense, thanks for the explanation. I'm happy to change to either or just leave it as-is, what is your preferred solution here?

petrochenkov replied:
Let's just use `tt` as the recovery for all erroneous fragment kinds.

tgross35 committed:
mbe: Use `TokenTree` as the fallback for invalid fragment specifiers

`tt` should match more, so use this for both missing and invalid
fragment specifiers.

Also remove one unneeded instance of `String`.

bors commented:
:pushpin: Commit cd5de49eaa522ee10c0e8a6b36a067791270dca2 has been approved by `petrochenkov`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=petrochenkov cd5de49eaa522ee10c0e8a6b36a067791270dca2 -->
<!-- homu: {"type":"Approved","sha":"cd5de49eaa522ee10c0e8a6b36a067791270dca2","approver":"petrochenkov","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
